### DISCUSSION TITLE
evaluete function and expression at compile time and buildin Regex

### DISCUSSION THREAD

**Discussion by SLusenti** [vlang/v#11726] [Ideas]
hi,
I didn't found anything like this in the documentation so i wrote here. Hope I'm not wrong.
I would like to implement something like  "comptime" ziglang statement.
so that if a variable or a function is marked as "at compile time" automatically that func is evaluated before compiling and then the returned value replace the func call (this could be done only if each of the element used in the function is known at build time).
this job could be done also with variable: instead of evaluate the variable expression at runtime, the variable is evaluated at build time and then the var value replace the var name reference in every occurrences (a sort of c #DEFINE macro).
this could lead at more optimized code.

For the syntax implementation I think we could use something very simple: if a var name or func name start/end with a special character ( I.E: underscore character  "_" ) then it's marked as "at compile time".

```v
_copiletime := 5* 100

//this will substitute 500 to _comptime reference before compiling
mut runtime := _compiletime * 20 

fn _compfn(_m int) int {
   return _m * 2
}

//this will substitute 1000  to _comptime reference before compiling
runtime = _compfn(_copiletime) * 20 
```

this feature could also lead at more readable code and some sweet code I.E perl like regex , like in the example below. 
(I'm not referring to a specific regex library)

```v
h := "hello world"


match (h) {
  // this will be replaced with regex.compile('hello worl\w').match(h ) {...} before compiling
  /hello worl\w/  { println(h) } 
}

//this will replace with regex.compile('hello worl\w').substitute(h, 'hello v' ) before compiling
sub := s/${h}/hello worl\w/hello v/

//will print --> hello v
println(sub) 
```
indeed if possible the regex could be compiled at build time ...

**Comment by JalonSolov** [comment]
Take a look at https://github.com/vlang/v/blob/master/doc/docs.md#conditional-compilation and https://github.com/vlang/v/blob/master/doc/docs.md#compile-time-reflection

**Comment by SLusenti** [comment]
i read that part of documentation,
i think I'm requesting something different.
I'm not asking for a sort of differential compilation. 
I'm asking for an enhancement for the compiler and for a precompilation text processor.
The precompilation text processor transform some "sweet more readable code" in function call
and then replace the function call with the returned value at build-time.
so, in case of regex we have this situation:
```v
match (h) {
  // this will be replaced with regex.compile('hello worl\w').match(h ) {...} before compiling
  /hello worl\w/  { println(h) } 
}
```
but at the end the compiler will not found the `/hello worl\w/ ` regex, but the struct returned by the `regex.compile()` function. 
so we have a readable code and lighter runtime because we already prepare and compile the regex at build time.
obvious the regex is just an example of how we could archive more readable and fast code.
but imagine a different scenario where you have multiple complex calculation.
and at end at runtime you don't have to do each step from the very beginning but run just what need"
 E.G.  in this code `mut _var := _complexfunc() / _complexnumb()` at run time we have just  `mut _var = [floating number]` the division has already be done at build-time.
